### ISSUE TITLE
output-consistency: COUNT hides evaluation errors

### ISSUE THREAD

**Issue by def-** [MaterializeInc/materialize#19662]
### What version of Materialize are you using?

1bcd1c4

### What is the issue?

```
DROP TABLE IF EXISTS dataflow_rendering;
CREATE TABLE dataflow_rendering (int8_neg_max INT8);
INSERT INTO dataflow_rendering VALUES (-9223372036854775808::INT8);
-- ------------------------------
-- Setup for evaluation strategy 'Constant folding':
CREATE OR REPLACE VIEW constant_folding (int8_neg_max)
 AS SELECT -9223372036854775808::INT8;
-- ------------------------------
-- Query using evaluation strategy 'Dataflow rendering':
SELECT
  count(trunc(acos(int8_neg_max)))
FROM
  dataflow_rendering;
-- ------------------------------
-- Query using evaluation strategy 'Constant folding':
SELECT
  count(trunc(acos(int8_neg_max)))
FROM
  constant_folding;
```
Via nightly, found using @nrainer-materialize 's new output-consistency test framework, which is meant to find just such differences: https://buildkite.com/materialize/nightlies/builds/2512#0188794e-3a88-43d6-9a9d-26cfda4627aa

The `count` in constant folding is ignoring that `acos` failed`, but in dataflow rendering the error is raised.

I think this is another instance of https://github.com/MaterializeInc/materialize/issues/17189 @ggevay @vmarcos Can you confirm?

**Comment by nrainer-materialize** [comment]
Some more context:

* `SELECT acos(int8_neg_max) FROM dataflow_rendering;` and `SELECT acos(int8_neg_max) FROM constant_folding;` fail with the same error message: `function acos is defined for numbers between -1 and 1 inclusive`
* The same applies to `trunc(acos(int8_neg_max))`.
* Applying `count` on the outcome succeeds fails with data-flow rendering but succeeds with constant folding.
* The table / view contains exactly one row.

```
materialize=> SELECT count(trunc(acos(int8_neg_max))) FROM dataflow_rendering;
ERROR:  Evaluation error: function acos is defined for numbers between -1 and 1 inclusive

materialize=> SELECT count(trunc(acos(int8_neg_max))) FROM constant_folding;
 count 
-------
     1
(1 row)
```

It think it is might be a consequence of #17189 but let others decide this.

@def-: In any case, I will add this to the ignore filter (unless you are already on that).

**Comment by def-** [comment]
> @def-: In any case, I will add this to the ignore filter (unless you are already on that).

Go ahead

**Comment by ggevay** [comment]
> It think it is might be a consequence of https://github.com/MaterializeInc/materialize/issues/17189

Ooh, this one is tricky! It's not really constant folding's fault, but `ReduceElision`'s: It calls [this code](https://github.com/MaterializeInc/materialize/blob/eadc9edb9cfdaa10282a809940a1fe1786a24308/src/expr/src/relation/mod.rs#L2254-L2261).

**Comment by ggevay** [comment]
But I think we shouldn't classify this one as a bug. I think the current consensus is that evaluation order is undefined in SQL. So, in a sense, the situation here is analogous to AND deciding to not evaluate it's second argument when it already determined that the first argument is true: When evaluating `count` here, we decide to not evaluate its argument, because we know that it can’t be null. (What specifically happens is that the above-linked code inserts an IS NULL check, and then at some point `MirScalarExpr::reduce` simplifies that to false, I guess.)

Still, I'll add this one to https://github.com/MaterializeInc/materialize/issues/17189.

**Comment by ggevay** [comment]
I'm wondering whether maybe all the error/non-error inconsistencies should be exempted from the output-consistency test, i.e., fail the test only if the differing result are non-error. What do you think @vmarcos, @aalexandrov?

**Comment by ggevay** [comment]
Btw. Haskell (citing it here as the canonical example of a lazy language) behaves similarly when counting elements of a list:
```
ghci> let l = [undefined, undefined, undefined]
ghci> show l
"[*** Exception: Prelude.undefined
CallStack (from HasCallStack):
  error, called at libraries/base/GHC/Err.hs:75:14 in base:GHC.Err
  undefined, called at <interactive>:43:10 in interactive:Ghci9
ghci> length l
3
```

**Comment by nrainer-materialize** [comment]
> I'm wondering whether maybe all the error/non-error inconsistencies should be exempted from the output-consistency test, i.e., fail the test only if the differing results are both non-error. What do you think @vmarcos, @aalexandrov?

I will ignore all cases where an aggregation is involved, data-flow rendering fails, and constant folding succeeds (https://github.com/MaterializeInc/materialize/pull/19668).

**Comment by aalexandrov** [comment]
> I'm wondering whether maybe all the error/non-error inconsistencies should be exempted from the output-consistency test, i.e., fail the test only if the differing results are both non-error. What do you think @vmarcos, @aalexandrov?

Makes sense to me. I'm not sure that this will be only constrained to aggregations, but we can expand as we find more examples.

**Comment by ggevay** [comment]
Another instance here: https://github.com/MaterializeInc/materialize/pull/28144#issuecomment-2232764474
